### PR TITLE
Added new sections on experimentation and agile project management

### PR DESCRIPTION
**Pull request by omri374**
Adding documentation on ML experimentation and ML Agile project management.

### CODE REVIEW

File: ml-fundamentals/ml-project-management.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+2. User stories and spikes are usually estimated using [T-shirt sizes](../agile-development/sprint-planning/estimation/readme.md#t-shirt-sizes) or similar, and not in actual days/hours. See more [here](../agile-development/sprint-planning/estimation/readme.md) on story estimation.
+
+3. ML design sessions should be included in each sprint.
+
+## Notes on collaboration between ML team and software development team
+
+- The ML and Software Development teams work together on the project. The team uses one backlog and attend the same Agile ceremonies. In cases where the project has a large number of participants, we will divide into working groups, but still have the entire team join the Agile ceremonies.
+
+- If possible, feasibility study and initial model experimentation takes place before the operationalization work kicks off.
+
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-project-management.md:42]:
```suggestion
```

File: ml-fundamentals/ml-project-management.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+- The ML team and dev team both share the accountability for the MLOps solution.
+
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-project-management.md:44]:
```suggestion
```

File: ml-fundamentals/ml-project-management.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+- The ML model interface (API) is determined as early as possible, to allow the developers to consider its integration into the production pipeline.
+
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-project-management.md:46]:
```suggestion
```

File: ml-fundamentals/ml-project-management.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+- MLOps artifacts are developed with a continuous collaboration and review of the ML team, to ensure the appropriate approaches for experimentation and
+productization are used.
+
```
Comment (TessFerrandez) [inline ml-fundamentals/ml-project-management.md:49]:
```suggestion
```

**Review by TessFerrandez — APPROVED** [review]
Great contribution - approving with some suggestions

File: ml-fundamentals/README.md [hunk 5]
```diff
@@ -10,15 +10,23 @@ This guideline documents the Machine Learning (ML) practices in CSE. CSE works w
...
-For specific aspects of an ML project, refer to the guidelines for different [project phases](#ml-project-phases).
+## ML Project phases in CSE
 
-## Sections
+![Project flow](flow.png)
 
-### General
+* [Envisioning](../design-reviews/recipes/engagement-process.md#envisioning--architecture-design-session-ads): Initial problem understanding, customer goals and objectives. *(Link to engineering playbook)*
+* Feasibility study: Assess whether the problem in question is feasible to solve satisfactorily using ML with the available data.
+* [Game plan](../design-reviews/recipes/engagement-process.md#game-plan): Defining the problem, scope, objectives, definition of done and high level plan for the entire project. *(Link to engineering playbook)*
```
[the remark this answers is not in the document]
Comment (omri374) [reply ml-fundamentals/README.md:22]: Thanks, I agree. I are working internally on making focus on a generic ML project and not CSE's

[the remark this answers is not in the document]
Comment (TessFerrandez) [reply ml-fundamentals/README.md]: Looks great to me

[the remark this answers is not in the document]
Comment (biancafbmd) [reply ml-fundamentals/README.md]: Looks good to me too :)

File: ml-fundamentals/README.md [hunk 6]
```diff
@@ -10,15 +10,23 @@ This guideline documents the Machine Learning (ML) practices in CSE. CSE works w
...
 
-For specific aspects of an ML project, refer to the guidelines for different [project phases](#ml-project-phases).
+## ML Project phases
 
-## Sections
+![Project flow](flow.png)
 
-### General
+* Envisioning: Initial problem understanding, customer goals and objectives.
+* Feasibility study: Assess whether the problem in question is feasible to solve satisfactorily using ML with the available data.
```
Comment (cmayomsft) [inline ml-fundamentals/README.md:21]: Do we assume we already have the data here? Or, if we do not, do we identify what data we'd need and how we'd collect it in order to move forward?

File: ml-fundamentals/ml-experimentation.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
+# Model Experimentation
+
+Machine learning model experimentation involves uncertainty around the expected model results and future operationalization.
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:3]: For these all the pages, should we have a heading for this intro text? It seems to just fall under the ToC. Maybe "Overview"?

File: ml-fundamentals/ml-experimentation.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+To handle this uncertainty as much as possible, we propose a semi-structured process, balancing between engineering/research best practices and rapid model/data exploration.
+
+## Model experimentation goals
+
+1. **Performance**: Find the best performing solution
+2. **Operationalization**: Keep an eye towards production, making sure the operationalization is feasible
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:9]: Nit. "making sure that operationalization is feasible"

File: ml-fundamentals/ml-experimentation.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+3. **Code quality** Maintain code and artifacts quality
+4. **Reproducibility**: Keep research active by allowing experiment tracking and reproducibility
+5. **Collaboration**: Foster the collaboration and joint work of multiple people on the team
+
+## Model experimentation challenges
+
+- **Trial and error process**: Difficult to plan and estimate durations and capacity.
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:16]: Nit. The section above is bulleted, this section is numbered. Should be consistent. Also, the first bullet ends with ".", the others don't. Not sure which is right, but should be consistent.

File: ml-fundamentals/ml-experimentation.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+## Tools and guidelines for successful model experimentation
+
+- [Virtual environments](#virtual-environments)
+- [Source control and folder/package structure](#source-control-and-folder-package-structure)
+- [Experiment tracking](#experiment-tracking)
+- [Datasets and models abstractions](#datasets-and-models-abstractions)
+- [Model evaluation](#model-evaluation)
+
+### Virtual environments
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:37]: We should mention that virtual environments allow us to be consistent across our local dev envs as well as with compute resources. We should use the same virt evn "source" *.txt/*yml. Anywhere we can build from the same source allows us to remove errors due to inconsistency.

File: ml-fundamentals/ml-experimentation.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+[Source control](https://github.com/microsoft/code-with-engineering-playbook/tree/master/source-control) should be applied to allow collaboration, versioning, code reviews, traceability and backup. In data science projects, source control should be used for code, and the storing and versioning of other  artifacts (e.g. data, scientific literature) should be decided upon depending on the scenario.
+
+#### Folder structure and source control expected outcomes
+
+- Defined folder structure for all users to use, pushed to the repo.
+- [.gitignore](https://git-scm.com/docs/gitignore) file determining which folders should be synced with `git` and which should be kept locally. For example, [this one](https://github.com/drivendata/cookiecutter-data-science/blob/master/%7B%7B%20cookiecutter.repo_name%20%7D%7D/.gitignore).
+- Determine how notebooks are stored and versioned (e.g. [strip output from Jupyter notebooks](https://github.com/kynan/nbstripout))
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:74]: To the point above, we need to balance flexibility with rigor with regard to experiementation in notebooks vs. operationalization. We should have guidance here. Can I create notebooks for quick/dirty experimentation and check them into a private branch I can share? Should I have to pass code quality conventions (linting, unit testing, etc.) to PR that notebook into dev? I think yes to both. Would love to discuss.

File: ml-fundamentals/ml-experimentation.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+#### Types of experiment tracking frameworks
+
+Experiment tracking frameworks differ by the set of features they provide for collecting experiment metadata, and comparing and analysing experiments. In CSE, we mainly use [MLFlow](https://mlflow.org/) on [Databricks](https://databricks.com/product/managed-mlflow) or [Azure ML Experimentation](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-track-experiments). Note that some experiment tracking frameworks require a deployment, while others are SaaS.
+
+#### Experiment tracking outcomes
+
+1. Decide on an experiment tracking framework
+2. Ensure it is accessible to all users
+3. Document set-up on local environments
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:94]: There is huge value to providing guidance on setting up the dev inner loop.

We should create all environments (dev, qa, prod, etc.) from source/IaC so they are consistent.

On of those environments created from that IaC script is my local development resource group/resource that map to my branch so when I'm developing something in a branch I can work in an resoure group indpendently. If I need to blow it away I can do so via IaC scripts.

Like virtual environments, we should document that so the team can get set up quickly/consistently.

Not sure if it belongs up here or closer to virtual environments.

File: ml-fundamentals/ml-experimentation.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+#### Abstraction benefits
+
+- Collaboration
+- Code quality
+- Reproducibility
+- Operationalization
+- Model performance
+
+### Model evaluation
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:131]: This section is structured differently from the others (Topic, Topic Outcomes, Topic Benefits). Why is that?

File: ml-fundamentals/ml-experimentation.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+When deciding on the evaluation of the ML model/process, consider the following checklist:
```
Comment (cmayomsft) [inline ml-fundamentals/ml-experimentation.md:133]: Should this be defined in the Feasibility section and refered to here? Seems like it should be defined in one place (DRY) and refered to where consumed (here).

File: ml-fundamentals/ml-project-management.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
+# Agile development considerations for ML projects
+
+## Goals
```
Comment (cmayomsft) [inline ml-fundamentals/ml-project-management.md:3]: No intro/overview paragraph like other pages.

File: ml-fundamentals/ml-project-management.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+Within this framework, the team follows these Agile ceremonies:
+
+- [Backlog management](../agile-development/backlog-management/readme.md)
+- [Retrospectives](../agile-development/retrospectives/readme.md)
+- [Scrum of Scrums](../agile-development/scrum-of-scrums/readme.md) (where applicable)
+- [Sprint planning](../agile-development/sprint-planning/readme.md)
+- [Standups](../agile-development/stand-ups/readme.md)
+- [Working agreement](../agile-development/team-agreements/working-agreements/readme.md)
+
+### Notes on Agile process during exploration and experimentation
```
Comment (cmayomsft) [inline ml-fundamentals/ml-project-management.md:23]: In CSE projects, we strive to get to an end to end "hello world" project repo with automated testing and DevOps ASAP so we can focus on iterating on different workstreams independently. 

What is that for ML? How can I get all that set up quickly so it's easy for the DS/ML team to start creating notebooks/scripts/models quickly while still working on the rest of the system (data collection, the rest of the solution around a placeholder/interim model, etc.).

**Review by cmayomsft — APPROVED** [review]
Approved with minor nits to be addressed.

Also suggested some additional details that are probably best for the backlog.

File: ml-fundamentals/README.md (hunk 6 above)
Comment (omri374) [reply ml-fundamentals/README.md:21]: There will be a full section on feasibility study. During this study we first verify we have access to data. Do you think we should add clarification here as well?

File: ml-fundamentals/ml-experimentation.md (hunk 11 above)
Comment (omri374) [reply ml-fundamentals/ml-experimentation.md:74]: Good point. Let's have a discussion on more specific guidance.

File: ml-fundamentals/ml-experimentation.md (hunk 12 above)
Comment (omri374) [reply ml-fundamentals/ml-experimentation.md:94]: Great points. Do you think the IaC scripts guidance and multiple environments setup should belong here or should we discuss this on a new page dealing with MLOps fundamentals, or maybe the ml-proposed-process doc which is coming soon.

File: ml-fundamentals/ml-experimentation.md (hunk 13 above)
Comment (omri374) [reply ml-fundamentals/ml-experimentation.md:131]: Added benefits to fit the other topics

File: ml-fundamentals/ml-experimentation.md (hunk 14 above)
Comment (omri374) [reply ml-fundamentals/ml-experimentation.md:133]: I totally agree. The feasibility study doc is still WIP, we can definitely move it there once it's ready and refer to it here. I'll keep it here for now until the feasibility study doc is ready and we discussed how this part should fit into the content there. cc @sarahdettorre

File: ml-fundamentals/ml-project-management.md (hunk 15 above)
Comment (omri374) [reply ml-fundamentals/ml-project-management.md:3]: fixed

File: ml-fundamentals/ml-project-management.md (hunk 16 above)
Comment (omri374) [reply ml-fundamentals/ml-project-management.md:23]: I'm not sure we have a one-size-fits-all here. For the feasibility study a CI process would likely be an overkill. After we finished with feasibility study, the ml-experimentation doc describes how to set up the env and code to get started. However some people would still not want things like linting for example at this stage. Let's continue the discussion and see if we can add a general guidance here.

File: ml-fundamentals/ml-experimentation.md (hunk 12 above)
Comment (biancafbmd) [reply ml-fundamentals/ml-experimentation.md:98]: These are all great point and I think they deserve more discussion. I would lean towards them belonging to MLOps fundamentals, but let's get more thoughts on this.

**Pull request merged**